[Replicate] Add unit test with replicate param names

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102401

This attribute wasn't actually used in tests, add a test ensuring that
if replicate is used on top of FSDP, the replicated parameter names are as
expected.

TODO: there are a few ways to check if module is managed by composable API,
such as replicated param names for replicate, _get_module_state API,
_get_registry_api, etc. We should unify all composable APIs to check in a
unified way (filed an issue)

Differential Revision: [D46236377](https://our.internmc.facebook.com/intern/diff/D46236377/)

rohan-varma committed:
[Replicate] Add unit test with replicate param names

This attribute wasn't actually used in tests, add a test ensuring that
if replicate is used on top of FSDP, the replicated parameter names are as
expected.

TODO: there are a few ways to check if module is managed by composable API,
such as replicated param names for replicate, _get_module_state API,
_get_registry_api, etc. We should unify all composable APIs to check in a
unified way (filed an issue)

Differential Revision: [D46236377](https://our.internmc.facebook.com/intern/diff/D46236377/)

[ghstack-poisoned]

awgu reviewed (approved):
Stamp to unblock since this looks easy. Lint and unit test failures look real. I can take a deeper look by beginning of next week.

awgu commented on test/distributed/_composable/test_compose.py:
```diff
@@ -145,6 +147,33 @@ def test_checkpoint_fsdp_submodules_use_reentrant(self):
+        fully_shard(model.u2, policy=ModuleWrapPolicy({nn.Linear}))
+        # replicate the rest
+        replicate(model)
+        # Run fwd + bwd to initialize DDP
```
Curious: what does it mean to initialize DDP? Do you have to run backward?

rohan-varma replied:
In replicate, DDP is only initialized in the first forward. You don't have to run backward, will clarify that in the comment

awgu commented on test/distributed/_composable/test_compose.py:
```diff
@@ -145,6 +147,33 @@ def test_checkpoint_fsdp_submodules_use_reentrant(self):
+            if mod not in [model.u1, model.u2]
+        ]
+        replicated_param_names = [
+            f"{module_name}.{n}"
```
So these are _not_ all prefixed from the same root module?

rohan-varma replied:
For something like
```
class MyModel:
    def __init__(self):
        self.lin1 = ...
        self.lin2 = ....
        self.p = nn.Parameter(...)
```

and if we replicate the whole thing, the names we expect are like `[lin1.weight, lin2.weight, p]`. They are all prefixed from the same root, but the root is `MyModel` and the root prefix is just `""`.

awgu commented on torch/distributed/_composable/replicate.py:
```diff
@@ -8,6 +8,8 @@
 
 from .contract import _get_registry, contract
 
+_ROOT_MODULE_PREFIX = ""
```
Could you explain why we define this? Is this meant to be changed in the future? Will it ever be something other than the empty string?

rohan-varma replied:
it is mostly so I don't hardcode `""` a few different times in the recursive name generation logic.

rohan-varma committed:
Update on "[Replicate] Add unit test with replicate param names"


This attribute wasn't actually used in tests, add a test ensuring that
if replicate is used on top of FSDP, the replicated parameter names are as
expected.

TODO: there are a few ways to check if module is managed by composable API,
such as replicated param names for replicate, _get_module_state API,
_get_registry_api, etc. We should unify all composable APIs to check in a
unified way (filed an issue)

Differential Revision: [D46236377](https://our.internmc.facebook.com/intern/diff/D46236377/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI done"
